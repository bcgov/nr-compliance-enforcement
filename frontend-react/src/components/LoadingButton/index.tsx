import React from 'react';

import {
  Button,
  InlineLoading
} from '@carbon/react';

import './styles.css';

// The clickFn must provide a return to inform if the action
// was successful or not, so the loading component can be properly
// adjusted
interface ButtonProps {
  id: string,
  clickFn: Function,
  status: {
    loading: string,
    success: string,
    error: string
  },
  label: string,
  icon?: JSX.Element
}

const LoadingButton = ({
  id,
  clickFn,
  status,
  label,
  icon
}: ButtonProps) => {
  const [loading, setLoading] = React.useState<boolean>(false);
  const [loadDesc, setLoadDesc] = React.useState<string>(status.loading);
  const [success, setSuccess] = React.useState<string>('');

  const setLoadingStates = async () => {
    setLoading(true);
    setSuccess('active');

    if (await clickFn()) {
      setSuccess('finished');
      setLoadDesc(status.success);
    } else {
      setSuccess('error');
      setLoadDesc(status.error);
    }

    // Before resetting the value for the next action,
    // set a small timeout to provide the user visual feedback
    setTimeout(() => {
      setLoading(false);
      setLoadDesc(status.loading);
      setSuccess('active');
    }, 1500);
  };

  return (
    <div>
      {
        loading ? (
          <InlineLoading
            className="buttonMargin"
            description={loadDesc}
            status={success}
            id={`loading-${id}`}
            data-testid={`loading-${id}`}
          />
        ) : (
          <Button
            onClick={() => setLoadingStates()}
            size="md"
            renderIcon={icon}
            id={`button-${id}`}
            data-testid={`button-${id}`}
          >
            {label}
          </Button>
        )
      }
    </div>
  );
};

export default LoadingButton;
