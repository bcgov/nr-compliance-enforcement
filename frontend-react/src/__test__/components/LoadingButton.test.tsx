/* eslint-disable no-undef */
import { render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import LoadingButton from '../../components/LoadingButton';

describe('the LoadingButton component', () => {
  const loadingStatus = {
    loading: 'Loading...',
    success: 'Finished!',
    error: 'Error'
  };

  it('should have the correct name', () => {
    const { getByTestId } = render(
      <LoadingButton
        id="test"
        clickFn={() => {}}
        label="Submit"
        status={loadingStatus}
      />
    );
    expect(getByTestId('button-test').textContent).toBe('Submit');
  });

  it('should match the snapshot', () => {
    const tree = renderer
      .create(
        <LoadingButton
          id="test"
          clickFn={() => {}}
          label="Submit"
          status={loadingStatus}
        />
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
