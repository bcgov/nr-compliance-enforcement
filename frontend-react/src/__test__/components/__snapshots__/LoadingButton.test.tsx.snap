// Vitest Snapshot v1

exports[`the LoadingButton component > should match the snapshot 1`] = `
<div>
  <button
    aria-describedby={null}
    aria-pressed={null}
    className="cds--btn cds--btn--md cds--btn--primary"
    data-testid="button-test"
    disabled={false}
    id="button-test"
    onClick={[Function]}
    tabIndex={0}
    type="button"
  >
    Submit
  </button>
</div>
`;
